File: 239-sliding-window-maximum/sliding-window-maximum.cpp
class Solution {
public:
    vector<int> maxSlidingWindow(vector<int>& nums, int k) {
        multiset<int>set;
        int n=nums.size(),s=0,e=0;
        vector<int>ans;
        while(e<n){
            set.insert(nums[e]);
            if(e-s+1==k){
                ans.push_back(*set.rbegin());
                set.erase(set.find(nums[s]));
                s++;
            }
            e++;
        }
        return ans;
    }
};

File: 239-sliding-window-maximum/sliding-window-maximum.java
class Solution {
    public int[] maxSlidingWindow(int[] nums, int k) {
        int n=nums.length;

        List<Integer> ls = new ArrayList<>();
        ArrayDeque<Integer> dq = new ArrayDeque<>();

        for(int i=0;i<n;i++){
            while(!dq.isEmpty() && dq.peekFirst()<=(i-k)){
                dq.pollFirst();
            }
            while(!dq.isEmpty() && nums[dq.peekLast()]<=nums[i]){
                dq.pollLast();
            }
            dq.offerLast(i);

            if(i+1>=k){
                ls.add(nums[dq.peekFirst()]);
            }
        }

        return ls.stream().mapToInt(num->(int)num).toArray();
    }
}